rust/gsc/src/gsc_tool/main.rs

```diff
@@ -0,0 +1,209 @@
+use inquire::Confirm;
+use migration::{DbErr, Migrator, MigratorTrait};
+use rand_core::{OsRng, RngCore};
+use sea_orm::{
+    ActiveModelTrait, ActiveValue, ColumnTrait, DatabaseConnection, EntityTrait, QueryFilter,
+};
+use std::io::{Error, ErrorKind};
+use tracing;
+
+mod table;
+
+#[tracing::instrument]
+async fn add_api_key(
+    opts: AddKeyOpts,
+    conn: &DatabaseConnection,
+) -> Result<(), Box<dyn std::error::Error>> {
+    // If the user hasn't specified a key generate one. This is
+    // the expected way for this tool to be used.
+    let key = match &opts.key {
+        None => {
+            let mut buf = [0u8; 24];
+            OsRng.fill_bytes(&mut buf);
+            BASE64.encode(&buf)
+        }
+        Some(key) => key.clone(),
+    };
+
+    // Go ahead and insert the key
+    let insert_key = api_key::ActiveModel {
+        id: ActiveValue::NotSet,
+        key: ActiveValue::Set(key.clone()),
+        desc: ActiveValue::Set(opts.desc.clone()),
+    };
+    tracing::info!("Adding key = {} as valid API key", &key);
+    insert_key.insert(conn).await?;
+
+    // If everything was a success go ahead and tell the user the
+    // API key.
+    tracing::info!("Key {} Added", &key);
+    println!("\nSuccessfully added {} as an API key", key);
+    Ok(())
+}
+
+#[tracing::instrument]
+async fn list_api_keys(_: ListKeysOpts, conn: &DatabaseConnection) -> Result<(), DbErr> {
+    let mut keys: Vec<_> = api_key::Entity::find()
+        .all(conn)
+        .await?
+        .into_iter()
+        .map(|x| {
+            vec![
+                format!("{}", x.id),
+                x.key,
+                textwrap::wrap(&x.desc, 60).join("\n"),
+            ]
+        })
+        .collect();
+
+    let headers = vec!["Id".into(), "Key".into(), "Desc".into()];
+    keys.insert(0, headers.clone());
+
+    table::print_table(keys);
+
+    Ok(())
+}
+
+#[tracing::instrument]
+async fn remove_api_key(
+    key: String,
+    conn: &DatabaseConnection,
+) -> Result<(), Box<dyn std::error::Error>> {
+    // First find the key, because its a very strange user expierence to see
+    // "key X was successfully removed" when it never existed in the first place.
+    // It means that you can never be quite sure you acomplished what you thought
+    // you did and have to use list-keys to check. This also lets us output
+    // the description before commiting.
+    let results = api_key::Entity::find()
+        .filter(api_key::Column::Key.eq(&key))
+        .all(conn)
+        .await?;
+
+    if results.len() == 0 {
+        println!("{} is not a valid key", key);
+        std::process::exit(2);
+    }
+
+    // We only want to prompt the user if the user can type into the terminal
+    if atty::is(Stream::Stdin) {
+        let should_delete = Confirm::new("Are you sure you want to delete this key?")
+            .with_default(false)
+            .with_help_message(format!("key = {}, desc = {:?}", key, results[0].desc).as_str())
```

Comment:
`results[0]` makes an implicit dependence on the fact that the key is unique in the DB. That's probably fine but if that ever changes then this code would delete more keys than the user expects.

May be  good to add another check fore `results.len() == 1`